compiler/rustc_mir_transform/src/reveal_all.rs

```diff
@@ -34,19 +34,19 @@ impl<'tcx> MutVisitor<'tcx> for RevealAllVisitor<'tcx> {
         _location: Location,
     ) {
         // Performance optimization: don't reintern if there is no `OpaqueCast` to remove.
-        if place.projection.iter().all(|elem| !matches!(elem, ProjectionElem::OpaqueCast(_))) {
-            return;
+        if place.projection.iter().any(|elem| matches!(elem, ProjectionElem::OpaqueCast(_))) {
+            // `OpaqueCast` projections are only needed if there are opaque types on which projections
+            // are performed. After the `RevealAll` pass, all opaque types are replaced with their
+            // hidden types, so we don't need these projections anymore.
+            place.projection = self.tcx.mk_place_elems(
+                &place
+                    .projection
+                    .into_iter()
+                    .filter(|elem| !matches!(elem, ProjectionElem::OpaqueCast(_)))
+                    .collect::<Vec<_>>(),
+            );
         }
-        // `OpaqueCast` projections are only needed if there are opaque types on which projections
-        // are performed. After the `RevealAll` pass, all opaque types are replaced with their
-        // hidden types, so we don't need these projections anymore.
-        place.projection = self.tcx.mk_place_elems(
-            &place
-                .projection
-                .into_iter()
-                .filter(|elem| !matches!(elem, ProjectionElem::OpaqueCast(_)))
-                .collect::<Vec<_>>(),
-        );
+
         self.super_place(place, _context, _location);
```

Comment:
while I don't fully understand in which cases we don't emit an `OpaqueCast` even though the type of a projection contains an opaque, it apparently can happen :thinking:

Reply:
Yea there are bugs here. The system is fragile and I haven't been able to come up with sth better, so without redoing projections from scratch, it may be best to just forbid projecting into opaques. I wanted to so that originally, but there was pushback saying some stuff should compile